File: Python/GameController.py
import NuclearReactor
from Bot import Bot
from Constants import Constants
import random
import copy
class GameController:
    def __init__(self,userbot,compbot):
        self.__userbot = userbot
        self.__compbot = compbot
        self.__cb = NuclearReactor.NuclearReactor()
        
        self.__currentPlayer = Constants.PlayerType.USER if(self.__toss()) else Constants.PlayerType.COMP

    def __toss(self):
        return random.randint(0,1)

    def play(self):
        while(self.__cb.getRecentOpponentMove().getStatus() == Constants.NuclearCellStatus.EMPTY or self.__cb.isGameOver() == Constants.PlayerType.NOPLAYER):
            if(self.__currentPlayer == Constants.PlayerType.USER):
                self.__cb.dropBomb(self.__userbot.calculateNextMove(copy.copy(self.__cb)),Constants.PlayerType.USER)
                self.__currentPlayer = Constants.PlayerType.COMP
            else:
                self.__cb.dropBomb(self.__compbot.calculateNextMove(copy.copy(self.__cb)),Constants.PlayerType.COMP)
                self.__currentPlayer = Constants.PlayerType.USER
        NuclearReactor.writeToJson((Constants.compbotname if(self.__currentPlayer) else Constants.userbotname))
    

File: C++/src/GameController.cpp
#include "GameController.h"


void GameController::play(){
    //cout << "Start Play" << endl;
    while(cb->getRecentOpponentMove().getStatus() == Constants::EMPTY || cb->isGameOver() == Constants::NOPLAYER){
        //cout << "turn " << endl;
        if(currentPlayer == Constants::USER){
            cb->dropBomb(userbot->calculateNextMove(*cb),Constants::USER);
            currentPlayer = Constants::COMP;
        }
        else{
            cb->dropBomb(compbot->calculateNextMove(*cb),Constants::COMP);
            currentPlayer = Constants::USER;
        }
    }
    writeToJson(((currentPlayer)?Constants::userbotname:Constants::compbotname));
}

//initial toss to decide which bot plays first
bool GameController::toss(){
    srand(time(NULL));
    return rand()%2;
}


GameController::GameController(Bot * userbot,Bot * compbot){
    //cout << "Game Controller Created" << endl;
    this->userbot = userbot;
    this->compbot = compbot;
    if(toss()){
        currentPlayer = Constants::USER;
    }
    else{
        currentPlayer = Constants::COMP;
    }
    cb = new NuclearReactor(currentPlayer);
}
